File: ndtv/R/ndtvAnimationWidget.R
#  File R/ndtvAnimationWidget.R in package ndtv, part of the Statnet suite
#  of packages for network analysis, http://statnet.org .
#
#  This software is distributed under the GPL-3 license.  It is free,
#  open source, and has the attribution requirements (GPL Section 7) at
#  http://statnet.org/attribution
#
#  Copyright 2016 Statnet Commons
#######################################################################

ndtvAnimationWidget <- function(out,options, width = NULL, height = NULL) {
  
  message('loading ndtv-d3 animation widget...')

  # forward options using x
  x = list(
    graphData = out,  # the network object and the rendering position data
    # the control options for ndtv-d3 need to be seperately JSON encoded or they will be mangled 
    # into a format that ndtv-d3 won't recognize
    animationOptions = jsonlite::toJSON(options,auto_unbox=TRUE,matrix='rowmajor', na = 'null',null='list')  
    
  )
  
  # the htmlwidgets package uses non-standard arguments to toJSON command
  # have to set these back to defaults for ndtv_d3 to parse correctly
  # via hack outlined in http://www.htmlwidgets.org/develop_advanced.html
  attr(x, 'TOJSON_ARGS') <- list(matrix='rowmajor', na = 'null',auto_unbox=FALSE,null='list')
  
  # create widget
  htmlwidgets::createWidget(
    name = 'ndtvAnimationWidget',
    x,
    width = width,
    height = height,
    package = 'ndtv',
    sizingPolicy = htmlwidgets::sizingPolicy(
      padding = 0,
      #defaultWidth = 500,
      #defaultHeight = 500
      #browser.fill = TRUE
    )
  )
}

#' Widget output function for use in Shiny
#'
#' @export
ndtvAnimationWidgetOutput <- function(outputId, width = '100%', height = '500px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'ndtvAnimationWidget', width, height, package = 'ndtv')
}

#' Widget render function for use in Shiny
#'
#' @export
renderNdtvAnimationWidget <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, ndtvAnimationWidgetOutput, env, quoted = TRUE)
}

File: ndtv/inst/htmlwidgets/ndtvAnimationWidget.js
HTMLWidgets.widget({

  name: 'ndtvAnimationWidget',

  type: 'output',
  

  initialize: function(el, width, height) {
    // get the DOM id of parent widget so can use it to construct a unique id for the animation
    ndtvId=d3.select(el).attr('id')+'_ndtvAnimation';
    // check if the ndtvAnimation div already exists as a child(if this is a re-render)
    if (!d3.select('#'+ndtvId).empty()){
      // remove its contents
      d3.select('#'+ndtvId).selectAll("*").remove();
    } else {
      // append the div into the DOM that ndtv_d3 will renderinto
      d3.select(el).append('div').attr('id',ndtvId).style('width',width+'px').style('height',height+'px');
    }
    return{};
  },

  renderValue: function(el, x, instance) {
      // get the DOM id of parent widget so can use it to construct a unique id for the animation
      ndtvId=d3.select(el).attr('id')+'_ndtvAnimation';
      // remove previous contents of the div (re render)
      d3.select('#'+ndtvId).selectAll("*").remove();
      // copy the graph options into the options array
      var options = x.animationOptions;
      options.graphData = x.graphData;
      // create the new ndtv instance (this will load the svg into the div with id ndtvId)
      graph = new ndtv_d3(options, '#'+ndtvId);
      return(graph);
  },

  resize: function(el, width, height, instance) {
    // get the DOM id of parent widget so can use it to construct a unique id for the animation
    ndtvId=d3.select(el).attr('id')+'_ndtvAnimation';
    d3.select('#'+ndtvId).style('width',width+'px').style('height',height+'px');
    graph.resizeGraph();  

  }

});
